### ISSUE TITLE
/data not decrypted on encrypted OnePlus One

### ISSUE THREAD

**Issue by NYZack** [TeamWin/Team-Win-Recovery-Project#291]
TWRP 2.7.1 doesn't give me the option to decrypt my encrypted OnePlus One when it starts up, so I can't access my data partition from within TWRP:  "Could not mount /data and unable to find crypto footer".  The phone boots just fine, however.

**Comment by jmonk** [comment]
experiencing same issue also. OnePlus One with TWRP 2.7.1.1.  Using Paranoid 4.5Beta1 full device encryption. I am able to boot fine to Paranoid but cannot access/mount data from TWRP.

**Comment by NYZack** [comment]
I was able to get my encrypted CM11 nightly to decrypt on TWRP 2.7.1.1 without a problem.  It was a recent fix in CM (July 27, I believe); maybe it hasn't been fixed in Paranoid.  Here's the description of the fix:
bacon: Relocate disk encryption metadata
Due to an issue outside our control, we cannot put the crypto
metadata in the userdata footer.
Move it to reserve4 and enable migration code.

**Comment by jmonk** [comment]
@NYZack can you share the CM11 nightlies that work?  I restored to factory(via fastboot and CM11s factory) and flash every nightlies available on cyanogenmod site and tried the encryption/decrypt but TWRP 2.7.1.1 seem to unable to decrypt the userdata.

**Comment by NYZack** [comment]
I'm currently on the 20140801 nightly, which works pretty flawlessly and can be decrypted in TWRP 2.7.1.1.

One thought:  when I start up TWRP, it always initially says that the data partition wasn't successfully decrypted, but then it does decrypt it.  Are you really sure it is not decrypting in your case, @jmonk?  (For example, go to the Advanced menu, open the File Explorer, and see if you can open the /data folder and see anything.)  My guess (and it's only a guess) is that TWRP looks for the crypto footer in the usual location, fails to find it, generates the error message, but then looks in the new location, finds it, and succesfully decrypts.

**Comment by Dees-Troy** [comment]
One of the troubles that I have with supporting decryption on the 1+1 right now is that the stable build that's shipping on the device is storing the decryption key in the footer -- a small empty space left at the end of the data partition. However, Cyngn Inc, in their infinite wisdom, has moved the decryption key in their nightlies to a special reserve4 partition. I'm not sure where other ROMs are storing the key right now, but basically you can thank Cyngn Inc for this clusterf*ck.

I think TWRP is still looking for the key in the footer, for now.

**Comment by jmonk** [comment]
@Dees-Troy thanks for the information and all your work.  Yes i can confirmed that TWRP is indeed looking the key in the footer of userdata.   I did found the bug report regarding what you have mentioned: https://jira.cyanogenmod.org/browse/BACON-98 I guess we would just need to wait until the new stable build properly migrated to reserve4 partition.  If not too much trouble, would it be possible to have a temporary build that is compatible with reserve4 partition?  

@NYZack i just tested with CM11 nightly (August 8, 2014 build) and TWRP 2.7.1.1 with no luck.  I ran the full device encryption, reboot the device and booted up to CM fine.  Rebooted the device to recovery and receive error on the terminal "could not mount /data and unable to find crypto footer".  I tried going to advance, filemanager, and browse /sdcard or /data.  No go, empty folder.

**Comment by NYZack** [comment]
@jmonk Weird - I don't know what to say.  Does TWRP give you the opportunity to enter the encryption password?  I just rebooted into TWRP, and it decrypted without a problem.  I don't want you to go on wild goose chase, but maybe a full wipe in TWRP followed by fresh installation of the CM11 nightly would help?  No idea, honestly.

**Comment by PaulPrime** [comment]
Is there anyone able to incorporate changes from http://review.cyanogenmod.org/#/c/68957/ to https://github.com/TeamWin/android_device_oneplus_bacon to be able to use decryption with O+O right now cause from what Steve Kondik said Cygn won't change this: 
"Steve Kondik added a comment - 26/Jul/14 4:45 PM
I wrote the migration piece (see the associated vold patch) to deal with this. It will just pick up the metadata from the old location and move it to the new place. TWRP should rebuild with this code and the new fstab."?

**Comment by NYZack** [comment]
It turns out that the only reason I was able to use TWRP to decrypt my /data partition was because I had shrunk the /data partition, applying the solution described in the (defunct) communications log pasted in the first comment in https://jira.cyanogenmod.org/plugins/servlet/mobile#issue/CYAN-87 (pasted below).  This worked beautifully, but I subsequently had problems dirty-flashing CM11 nightlies; I'm not sure if it was related.  Warning to anybody who tries this:  backing up and restoring by using tar created a lot of grief due to the size of the resultant tar'ed file.  I would suggest using TWRP's built-in backup and restore functionality, though I didn't try it.

Reboot into recovery (I'm using 4EXT Recovery, though CWM Recovery should work too)
Make sure /system, /sdcard, and /data are all mounted
From an adb shell, make a tar backup of /data (make sure you first have enough space on your SD card):
/system/xbin/tar -C / -cf /sdcard/data.tar data
Figure out device name of /data partition:
grep /data /proc/mounts
As an example, let's say the previous step gave you /dev/block/mmcblk0p23
Figure out the partition's size and the new size of the /data filesystem:
data_dev_size=$(cat /sys/class/block/mmcblk0p23/size)
new_fs_size=$(( ((data_dev_size/2)-16)*1024 ))
Reformat /data with correct size (continuing with the example of mmcblk0p23):
umount /data
/system/bin/make_ext4fs -a /data -l $new_fs_size /dev/block/mmcblk0p23
Remount /data
Restore the backup of /data:
/system/xbin/tar -C / -xf /sdcard/data.tar
Reboot

# The "Encrypt phone" option in the Settings app now properly encrypts /data on the phone.

**Comment by PaulPrime** [comment]
@NYZack the problem is with TWRP not being able to decrypt encrypted Oneplus One. Encryption works perfectly fine.

**Comment by jfeise** [comment]
Until TWRP makes the changes, it is of course possible, at least for people who build their own nightlies, to just revert Steve Kondik's changes, to keep the encryption key in the /data footer.
I just did that, and TWRP can decrypt the /data partition fine. That keeps me happy for now.
Ultimately, TWRP of course needs to read the key from the new location in reserve4.

**Comment by tempxxx** [comment]
I can confirm that the latest Cyanogenmod nightly + TWRP 2.7.1.1 does NOT decrypt the volume. This is a frustrating bug.  I had to move to Cyanogenmod nightlies just to get encryption working on the OnePlus, and now TWRP won't decrypt the partition, which causes the nightly update process to fail.  Ugh.  

Doesn't look like there are any commits to address this problem.  Can someone confirm that something is in the works?

Also, is it possible to manually upgrade each nightly without successfully decrypting the volumes within TWRP recovery?  Alternatively, are there any other recoveries that are capable of doing this?

**Comment by jfeise** [comment]
The "normal" nightly update process fails because it downloads the nightly to /sdcard, i.e., /data/media/0, and that can't be decrypted in recovery.
You can update the nightly manually by downloading it, and then doing an adb sideload. The sideload writes it to /tmp if /data isn't available. I used to do that when I used CWM, before I switched to TWRP.
This is safe because the nightlies only update /system, which isn't encrypted.

**Comment by tempxxx** [comment]
Decryption is somewhat working in TWRP 2.8, but there's a new issue introduced where it's reporting that the storage cannot be mounted and the Nightly update fails.  I get this:

E: Unable to mount storage.
[...]
Data successfully decrypted, new block device: '/dev/block/dm-0'
[...]
Can't install this package on top of incompatible data.  Please try another package or run a factory reset
E: Error executing updater binary in zip '/data/media/0//cmupdater//cm-11-20140910-NIGHTLY-bacon.zip'
E: Error installing zip file '/data/media/0//cmupdater//cm-11-20140910-NIGHTLY-bacon.zip'

**Comment by NYZack** [comment]
I can verify tempxxx's experience.  I was able to install a new nightly only by copying the update zip to /tmp, unmounting /data, and installing the zip file from /tmp (I had to make sure that the install directory was initially set to /tmp, since otherwise /data was remounted when I selected install).  However, this created a boatload of other problems:  the phone would hang for a long time on reboot after entering my password (with the screen blank and only the capacitive buttons lit), and the "Updating app..." screen never appeared.  When the phone finally rebooted, my Play Services kept crashing, and other apps were messed up.  My guess is that Dalvik Cache never got rebuilt.  Had to restore an old backup.

So still definitely not working properly.

**Comment by jfeise** [comment]
The blank screen is known. I reported that on the CM forum some time ago. Apparently, CM tries to load the resource before /data is decrypted.
It updates the apps, as adb logcat shows.
The failure to install updates is a TWRP bug, I think.

**Comment by NYZack** [comment]
@jfeise : Hmmm.  Can you direct me to your report on the CM forum?  I wonder if we're talking about the same blank screen issue.  Since encrypting, whenever I reboot, the screen is briefly blank after entering my password, but it's not a problem.  However, when I tried to install the latest (0712) nightly (the first time I've tried to install a nightly since encrypting this time around), the screen was blank for a very long time (a few minutes), and then the phone was unusable once it ultimately rebooted (mostly manifested by Play Services repeatedly force-closing).  It's possible it's related to the 0712 nightly, I suppose.

**Comment by jfeise** [comment]
http://forum.cyanogenmod.org/topic/99006-android-is-upgrading-dialog-missing-in-recent-nightlies/
It looks like the issue you are having, except that for me, everything works once it is done.

**Comment by jfeise** [comment]
The latest 2.8.0.1 TWRP update solves the unable to install issue.

**Comment by srt10coupe** [comment]
E: Error executing updater binary in zip 
same Error in 2.8.0.1 TWRP and other device (HTC One m7).

**Comment by tempxxx** [comment]
Confirming that 2.8.0.1 solves the issue me too

**Comment by master-sonic** [comment]
Decryption works again with build 38R (rollout began on 10/07/2014) and TWRP 2.8.0.1.

**Comment by raaomoka** [comment]
Hey guys, I know this is a very old thread, but I'm experiencing similar issue on my OnePlus One with TWRP v2.8.6.0 and CyanogenMod v12.1-YOG4PAS3JL. Is there a way to solve this?

I'm using TWRP v2.8.6.0 'cause apparently, later versions don't update my baseband for whatever reason when I flash a later ROM...

**Comment by jfeise** [comment]
Use TWRP 2.8.7.0.
File a bug report about the baseband issue.

**Comment by raaomoka** [comment]
Updated to v2.8.7.0 and yeah it's working fine with this CyanogenMod build, thanks. Here's the report on the baseband not updating, still ongoing:

https://github.com/TeamWin/Team-Win-Recovery-Project/issues/444

**Comment by berrythesoftwarecodeprogrammar** [comment]
im having this issue on my galaxy nexus (maguro) with twrp 2.8.7.0. my system is encrypted and boots fine but twrp wont decrypt it or ask to decrypt so i cant mount anything. using cyanogenmod 11. i want to upgrade to cyanogenmod 12.1 but i cant use twrp..

**Comment by berrythesoftwarecodeprogrammar** [comment]
this person seems to have fixed something but its only for a certain phone http://www.teamsuperluminal.org/forums/topic/recovery-and-encrypted-data-partition/

**Comment by jfeise** [comment]
Can we please keep the discussion on topic? This is a thread for the OnePlus One, not Galaxy Nexus or Xiaomi or any other phone. Please create separate issues for your phones.
TWRP 3.0.2 decrypts /data fine for the OnePlus One. (3.0.1 had an issue with the password entry field that resulted in decryption not working at all, see the release notes for 3.0.2.)

**Comment by antoszka** [comment]
Unfortunately, I cannot confirm the most recent comment by @jfeise ("TWRP 3.0.2 decrypts /data fine"), I could not get it to decrypt my data partition no matter which type of lock/password I chose, nor whether I had the switch "require password to boot" on or off. I managed to update my system (from the March SNAPSHOT to the April SNAPSHOT) by booting into trwp-3.0.2-0, skipping /data decryption, mounting an OTG stick and selecting the update from there, but the problem most certainly still exists.

**Comment by couto** [comment]
In my experience (bacon with TWRP 3.0.2 and CyanogenMod 13.0-20160820-NIGHTLY) both @jfeise and @antoszka are actually right.

I can decrypt /data if I have a 3x3 pattern, but a 6x6 pattern won't do.
